server/src/main/java/org/elasticsearch/common/geo/GenericPointParser.java

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.geo;
+
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentSubParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.Function;
+
+import static org.elasticsearch.common.Strings.collectionToDelimitedString;
+
+/**
+ * This class can parse points from XContentParser and supports several formats:
+ * @param <T> type of point to produce (either geographic or cartesian)
+ */
+public abstract class GenericPointParser<T> {
+    private static final String X_FIELD = "x";
+    private static final String Y_FIELD = "y";
+    private static final String Z_FIELD = "z";
+    private static final String GEOHASH = "geohash";
```

Comment:
geohash only makes sense for geo which makes it difficult to have it as a generic?

Reply:
see https://github.com/elastic/elasticsearch/pull/85442#discussion_r855104508